CI is not happy about this:
```
Error: ]: Build step io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor#setupEndpoints threw an exception: java.lang.RuntimeException: java.lang.RuntimeException: Failed to process method 'io.quarkus.resteasy.reactive.server.test.customproviders.UniFiltersResource#filters'
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:278)
	at io.quarkus.resteasy.reactive.server.deployment.ResteasyReactiveProcessor.setupEndpoints(ResteasyReactiveProcessor.java:525)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:882)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.lang.RuntimeException: Failed to process method 'io.quarkus.resteasy.reactive.server.test.customproviders.UniFiltersResource#filters'
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod(EndpointIndexer.java:639)
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:366)
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createEndpoints(EndpointIndexer.java:249)
	... 12 more
Caused by: java.lang.NullPointerException
	at io.quarkus.resteasy.reactive.common.deployment.JsonDefaultProducersHandler.handle(JsonDefaultProducersHandler.java:26)
	at io.quarkus.resteasy.reactive.server.deployment.QuarkusServerEndpointIndexer.isDefaultJson(QuarkusServerEndpointIndexer.java:171)
	at io.quarkus.resteasy.reactive.server.deployment.QuarkusServerEndpointIndexer.warnAboutMissingJsonProviderIfNeeded(QuarkusServerEndpointIndexer.java:163)
	at io.quarkus.resteasy.reactive.server.deployment.QuarkusServerEndpointIndexer.handleAdditionalMethodProcessing(QuarkusServerEndpointIndexer.java:156)
	at io.quarkus.resteasy.reactive.server.deployment.QuarkusServerEndpointIndexer.handleAdditionalMethodProcessing(QuarkusServerEndpointIndexer.java:32)
	at org.jboss.resteasy.reactive.common.processor.EndpointIndexer.createResourceMethod(EndpointIndexer.java:631)
	... 14 more

	at io.quarkus.test.QuarkusUnitTest.beforeAll(QuarkusUnitTest.java:613)
```

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/QuarkusServerEndpointIndexer.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/QuarkusServerEndpointIndexer.java
@@ -7,44 +7,60 @@ import java.util.function.Predicate;
 import javax.ws.rs.core.MediaType;
 
 import org.jboss.jandex.AnnotationInstance;
+import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
+import org.jboss.jandex.MethodInfo;
 import org.jboss.jandex.Type;
+import org.jboss.logging.Logger;
 import org.jboss.resteasy.reactive.common.ResteasyReactiveConfig;
 import org.jboss.resteasy.reactive.common.processor.DefaultProducesHandler;
+import org.jboss.resteasy.reactive.common.processor.transformation.AnnotationStore;
+import org.jboss.resteasy.reactive.server.model.ServerResourceMethod;
 import org.jboss.resteasy.reactive.server.processor.ServerEndpointIndexer;
 import org.jboss.resteasy.reactive.server.processor.ServerIndexedParameter;
 import org.jboss.resteasy.reactive.server.spi.EndpointInvokerFactory;
 
+import io.quarkus.deployment.Capabilities;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
 import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
 import io.quarkus.deployment.builditem.nativeimage.ReflectiveClassBuildItem;
+import io.quarkus.resteasy.reactive.common.deployment.JsonDefaultProducersHandler;
 import io.quarkus.resteasy.reactive.server.runtime.ResteasyReactiveRecorder;
 
 public class QuarkusServerEndpointIndexer
         extends ServerEndpointIndexer {
+
+    private static final org.jboss.logging.Logger LOGGER = Logger.getLogger(QuarkusServerEndpointIndexer.class);
+
+    private final Capabilities capabilities;
     private final BuildProducer<GeneratedClassBuildItem> generatedClassBuildItemBuildProducer;
     private final BuildProducer<BytecodeTransformerBuildItem> bytecodeTransformerBuildProducer;
     private final BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer;
     private final DefaultProducesHandler defaultProducesHandler;
+    private final JsonDefaultProducersHandler jsonDefaultProducersHandler;
     private final ResteasyReactiveRecorder resteasyReactiveRecorder;
 
     private final Predicate<String> applicationClassPredicate;
 
     QuarkusServerEndpointIndexer(Builder builder) {
         super(builder);
+        this.capabilities = builder.capabilities;
         this.generatedClassBuildItemBuildProducer = builder.generatedClassBuildItemBuildProducer;
         this.bytecodeTransformerBuildProducer = builder.bytecodeTransformerBuildProducer;
         this.reflectiveClassProducer = builder.reflectiveClassProducer;
         this.defaultProducesHandler = builder.defaultProducesHandler;
         this.applicationClassPredicate = builder.applicationClassPredicate;
         this.resteasyReactiveRecorder = builder.resteasyReactiveRecorder;
+        this.jsonDefaultProducersHandler = new JsonDefaultProducersHandler();
     }
 
+    private DefaultProducesHandler.Context currentDefaultProducesContext;
+
     @Override
-    protected String[] applyAdditionalDefaults(Type nonAsyncReturnType) {
-        List<MediaType> defaultMediaTypes = defaultProducesHandler.handle(new DefaultProducesHandler.Context() {
+    protected void setupApplyDefaults(Type nonAsyncReturnType) {
+        currentDefaultProducesContext = new DefaultProducesHandler.Context() {
             @Override
             public Type nonAsyncReturnType() {
                 return nonAsyncReturnType;
@@ -59,7 +75,12 @@ public class QuarkusServerEndpointIndexer
             public ResteasyReactiveConfig config() {
                 return config;
             }
-        });
+        };
+    }
+
+    @Override
+    protected String[] applyAdditionalDefaults(Type nonAsyncReturnType) {
+        List<MediaType> defaultMediaTypes = defaultProducesHandler.handle(currentDefaultProducesContext);
         if ((defaultMediaTypes != null) && !defaultMediaTypes.isEmpty()) {
             String[] result = new String[defaultMediaTypes.size()];
             for (int i = 0; i < defaultMediaTypes.size(); i++) {
@@ -80,12 +101,18 @@ public class QuarkusServerEndpointIndexer
 
     public static final class Builder extends AbstractBuilder<Builder> {
 
+        private final Capabilities capabilities;
+
         private BuildProducer<GeneratedClassBuildItem> generatedClassBuildItemBuildProducer;
         private BuildProducer<BytecodeTransformerBuildItem> bytecodeTransformerBuildProducer;
         private BuildProducer<ReflectiveClassBuildItem> reflectiveClassProducer;
         private ResteasyReactiveRecorder resteasyReactiveRecorder;
         private DefaultProducesHandler defaultProducesHandler = DefaultProducesHandler.Noop.INSTANCE;
         public Predicate<String> applicationClassPredicate;
+
+        public Builder(Capabilities capabilities) {
+            this.capabilities = capabilities;
+        }
 
         @Override
         public QuarkusServerEndpointIndexer build() {
@@ -124,5 +151,56 @@ public class QuarkusServerEndpointIndexer
             this.defaultProducesHandler = defaultProducesHandler;
             return this;
         }
+    }
+
+    @Override
+    protected void handleAdditionalMethodProcessing(ServerResourceMethod method, ClassInfo currentClassInfo,
+            MethodInfo info, AnnotationStore annotationStore) {
+        super.handleAdditionalMethodProcessing(method, currentClassInfo, info, annotationStore);
+        warnAboutMissingJsonProviderIfNeeded(method, info);
+    }
+
+    private void warnAboutMissingJsonProviderIfNeeded(ServerResourceMethod method, MethodInfo info) {
+        if (!capabilities.isCapabilityWithPrefixMissing("io.quarkus.resteasy.reactive.json")) {
+            return;
+        }
+        if (hasJson(method) || isDefaultJson()) {
+            LOGGER.warnf("Quarkus detected the use of JSON in JAX-RS method '" + info.declaringClass().name() + "#"
+                    + info.name()
+                    + "' but no JSON extension has been added. Consider adding 'quarkus-resteasy-reactive-jackson' or 'quarkus-resteasy-reactive-jsonb'.");
+        }
+    }
+
+    private boolean isDefaultJson() {
+        List<MediaType> mediaTypes = jsonDefaultProducersHandler.handle(currentDefaultProducesContext);
+        for (MediaType mediaType : mediaTypes) {
+            if (isJson(mediaType.toString())) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean hasJson(ServerResourceMethod method) {
+        return hasJson(method.getProduces()) || hasJson(method.getConsumes()) || isJson(method.getSseElementType());
+    }
+
+    private boolean hasJson(String[] types) {
+        if (types == null) {
+            return false;
+        }
+        for (String type : types) {
+            if (isJson(type)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private boolean isJson(String type) {
+        if (type == null) {
+            return false;
+        }
+        return type.startsWith(MediaType.APPLICATION_JSON);
     }
 }

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveProcessor.java
@@ -339,7 +339,8 @@ public class ResteasyReactiveProcessor {
             ParamConverterProvidersBuildItem paramConverterProvidersBuildItem,
             List<ApplicationClassPredicateBuildItem> applicationClassPredicateBuildItems,
             List<MethodScannerBuildItem> methodScanners,
-            List<AnnotationsTransformerBuildItem> annotationTransformerBuildItems)
+            List<AnnotationsTransformerBuildItem> annotationTransformerBuildItems,
+            Capabilities capabilities)
             throws NoSuchMethodException {
 
         if (!resourceScanningResultBuildItem.isPresent()) {
@@ -390,77 +391,68 @@ public class ResteasyReactiveProcessor {
             BiConsumer<String, BiFunction<String, ClassVisitor, ClassVisitor>> transformationConsumer = (name,
                     function) -> bytecodeTransformerBuildItemBuildProducer
                             .produce(new BytecodeTransformerBuildItem(name, function));
-            QuarkusServerEndpointIndexer.Builder serverEndpointIndexerBuilder = new QuarkusServerEndpointIndexer.Builder()
-                    .addMethodScanners(
-                            methodScanners.stream().map(MethodScannerBuildItem::getMethodScanner).collect(toList()))
-                    .setIndex(index)
-                    .addContextTypes(CONTEXT_TYPES)
-                    .setFactoryCreator(new QuarkusFactoryCreator(recorder, beanContainerBuildItem.getValue()))
-                    .setEndpointInvokerFactory(new QuarkusInvokerFactory(generatedClassBuildItemBuildProducer, recorder))
-                    .setGeneratedClassBuildItemBuildProducer(generatedClassBuildItemBuildProducer)
-                    .setBytecodeTransformerBuildProducer(bytecodeTransformerBuildItemBuildProducer)
-                    .setReflectiveClassProducer(reflectiveClassBuildItemBuildProducer)
-                    .setExistingConverters(existingConverters)
-                    .setScannedResourcePaths(scannedResourcePaths)
-                    .setConfig(createRestReactiveConfig(config))
-                    .setAdditionalReaders(additionalReaders)
-                    .setHttpAnnotationToMethod(result.getHttpAnnotationToMethod())
-                    .setInjectableBeans(injectableBeans)
-                    .setAdditionalWriters(additionalWriters)
-                    .setDefaultBlocking(appResult.getBlockingDefault())
-                    .setApplicationScanningResult(appResult)
-                    .setMultipartParameterIndexerExtension(
-                            new GeneratedMultipartParamIndexerExtension(transformationConsumer, classOutput))
-                    .setMultipartReturnTypeIndexerExtension(
-                            new GeneratedHandlerMultipartReturnTypeIndexerExtension(classOutput))
-                    .setFieldInjectionIndexerExtension(
-                            new TransformedFieldInjectionIndexerExtension(transformationConsumer, false, (field) -> {
-                                initConverters.invokeStaticMethod(MethodDescriptor.ofMethod(field.getInjectedClassName(),
-                                        field.getMethodName(),
-                                        void.class, Deployment.class),
-                                        initConverters.getMethodParam(0));
-                            }))
-                    .setConverterSupplierIndexerExtension(new GeneratedConverterIndexerExtension(
-                            (name) -> new GeneratedClassGizmoAdaptor(generatedClassBuildItemBuildProducer,
-                                    applicationClassPredicate.test(name))))
-                    .setHasRuntimeConverters(!paramConverterProviders.getParamConverterProviders().isEmpty())
-                    .setClassLevelExceptionMappers(
-                            classLevelExceptionMappers.isPresent() ? classLevelExceptionMappers.get().getMappers()
-                                    : Collections.emptyMap())
-                    .setResourceMethodCallback(new Consumer<>() {
-                        @Override
-                        public void accept(EndpointIndexer.ResourceMethodCallbackData entry) {
-                            MethodInfo method = entry.getMethodInfo();
+            QuarkusServerEndpointIndexer.Builder serverEndpointIndexerBuilder = new QuarkusServerEndpointIndexer.Builder(
+                    capabilities)
+                            .addMethodScanners(
+                                    methodScanners.stream().map(MethodScannerBuildItem::getMethodScanner).collect(toList()))
+                            .setIndex(index)
+                            .addContextTypes(CONTEXT_TYPES)
+                            .setFactoryCreator(new QuarkusFactoryCreator(recorder, beanContainerBuildItem.getValue()))
+                            .setEndpointInvokerFactory(
+                                    new QuarkusInvokerFactory(generatedClassBuildItemBuildProducer, recorder))
+                            .setGeneratedClassBuildItemBuildProducer(generatedClassBuildItemBuildProducer)
+                            .setBytecodeTransformerBuildProducer(bytecodeTransformerBuildItemBuildProducer)
+                            .setReflectiveClassProducer(reflectiveClassBuildItemBuildProducer)
+                            .setExistingConverters(existingConverters)
+                            .setScannedResourcePaths(scannedResourcePaths)
+                            .setConfig(createRestReactiveConfig(config))
+                            .setAdditionalReaders(additionalReaders)
+                            .setHttpAnnotationToMethod(result.getHttpAnnotationToMethod())
+                            .setInjectableBeans(injectableBeans)
+                            .setAdditionalWriters(additionalWriters)
+                            .setDefaultBlocking(appResult.getBlockingDefault())
+                            .setApplicationScanningResult(appResult)
+                            .setMultipartParameterIndexerExtension(
+                                    new GeneratedMultipartParamIndexerExtension(transformationConsumer, classOutput))
+                            .setMultipartReturnTypeIndexerExtension(
+                                    new GeneratedHandlerMultipartReturnTypeIndexerExtension(classOutput))
+                            .setFieldInjectionIndexerExtension(
+                                    new TransformedFieldInjectionIndexerExtension(transformationConsumer, false, (field) -> {
+                                        initConverters.invokeStaticMethod(
+                                                MethodDescriptor.ofMethod(field.getInjectedClassName(),
+                                                        field.getMethodName(),
+                                                        void.class, Deployment.class),
+                                                initConverters.getMethodParam(0));
+                                    }))
+                            .setConverterSupplierIndexerExtension(new GeneratedConverterIndexerExtension(
+                                    (name) -> new GeneratedClassGizmoAdaptor(generatedClassBuildItemBuildProducer,
+                                            applicationClassPredicate.test(name))))
+                            .setHasRuntimeConverters(!paramConverterProviders.getParamConverterProviders().isEmpty())
+                            .setClassLevelExceptionMappers(
+                                    classLevelExceptionMappers.isPresent() ? classLevelExceptionMappers.get().getMappers()
+                                            : Collections.emptyMap())
+                            .setResourceMethodCallback(new Consumer<>() {
+                                @Override
+                                public void accept(EndpointIndexer.ResourceMethodCallbackData entry) {
+                                    MethodInfo method = entry.getMethodInfo();
 
-                            resourceMethodEntries.add(new ResteasyReactiveResourceMethodEntriesBuildItem.Entry(
-                                    entry.getBasicResourceClassInfo(), method,
-                                    entry.getActualEndpointInfo(), entry.getResourceMethod()));
+                                    resourceMethodEntries.add(new ResteasyReactiveResourceMethodEntriesBuildItem.Entry(
+                                            entry.getBasicResourceClassInfo(), method,
+                                            entry.getActualEndpointInfo(), entry.getResourceMethod()));
 
-                            String source = ResteasyReactiveProcessor.class.getSimpleName() + " > " + method.declaringClass()
-                                    + "[" + method + "]";
+                                    String source = ResteasyReactiveProcessor.class.getSimpleName() + " > "
+                                            + method.declaringClass()
+                                            + "[" + method + "]";
 
-                            ClassInfo classInfoWithSecurity = consumeStandardSecurityAnnotations(method,
-                                    entry.getActualEndpointInfo(), index, c -> c);
-                            if (classInfoWithSecurity != null) {
-                                reflectiveClass.produce(new ReflectiveClassBuildItem(false, true, false,
-                                        entry.getActualEndpointInfo().name().toString()));
-                            }
+                                    ClassInfo classInfoWithSecurity = consumeStandardSecurityAnnotations(method,
+                                            entry.getActualEndpointInfo(), index, c -> c);
+                                    if (classInfoWithSecurity != null) {
+                                        reflectiveClass.produce(new ReflectiveClassBuildItem(false, true, false,
+                                                entry.getActualEndpointInfo().name().toString()));
+                                    }
 
-                            reflectiveHierarchy.produce(new ReflectiveHierarchyBuildItem.Builder()
-                                    .type(method.returnType())
-                                    .index(index)
-                                    .ignoreTypePredicate(QuarkusResteasyReactiveDotNames.IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
-                                    .ignoreFieldPredicate(QuarkusResteasyReactiveDotNames.IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
-                                    .ignoreMethodPredicate(
-                                            QuarkusResteasyReactiveDotNames.IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
-                                    .source(source)
-                                    .build());
-
-                            for (short i = 0; i < method.parameters().size(); i++) {
-                                Type parameterType = method.parameters().get(i);
-                                if (!hasAnnotation(method, i, ResteasyReactiveServerDotNames.CONTEXT)) {
                                     reflectiveHierarchy.produce(new ReflectiveHierarchyBuildItem.Builder()
-                                            .type(parameterType)
+                                            .type(method.returnType())
                                             .index(index)
                                             .ignoreTypePredicate(
                                                     QuarkusResteasyReactiveDotNames.IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
@@ -470,24 +462,39 @@ public class ResteasyReactiveProcessor {
                                                     QuarkusResteasyReactiveDotNames.IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
                                             .source(source)
                                             .build());
-                                }
-                            }
-                        }
 
-                        private boolean hasAnnotation(MethodInfo method, short paramPosition, DotName annotation) {
-                            for (AnnotationInstance annotationInstance : method.annotations()) {
-                                AnnotationTarget target = annotationInstance.target();
-                                if (target != null && target.kind() == AnnotationTarget.Kind.METHOD_PARAMETER
-                                        && target.asMethodParameter().position() == paramPosition
-                                        && annotationInstance.name().equals(annotation)) {
-                                    return true;
+                                    for (short i = 0; i < method.parameters().size(); i++) {
+                                        Type parameterType = method.parameters().get(i);
+                                        if (!hasAnnotation(method, i, ResteasyReactiveServerDotNames.CONTEXT)) {
+                                            reflectiveHierarchy.produce(new ReflectiveHierarchyBuildItem.Builder()
+                                                    .type(parameterType)
+                                                    .index(index)
+                                                    .ignoreTypePredicate(
+                                                            QuarkusResteasyReactiveDotNames.IGNORE_TYPE_FOR_REFLECTION_PREDICATE)
+                                                    .ignoreFieldPredicate(
+                                                            QuarkusResteasyReactiveDotNames.IGNORE_FIELD_FOR_REFLECTION_PREDICATE)
+                                                    .ignoreMethodPredicate(
+                                                            QuarkusResteasyReactiveDotNames.IGNORE_METHOD_FOR_REFLECTION_PREDICATE)
+                                                    .source(source)
+                                                    .build());
+                                        }
+                                    }
                                 }
-                            }
-                            return false;
-                        }
-                    })
-                    .setResteasyReactiveRecorder(recorder)
-                    .setApplicationClassPredicate(applicationClassPredicate);
+
+                                private boolean hasAnnotation(MethodInfo method, short paramPosition, DotName annotation) {
+                                    for (AnnotationInstance annotationInstance : method.annotations()) {
+                                        AnnotationTarget target = annotationInstance.target();
+                                        if (target != null && target.kind() == AnnotationTarget.Kind.METHOD_PARAMETER
+                                                && target.asMethodParameter().position() == paramPosition
+                                                && annotationInstance.name().equals(annotation)) {
+                                            return true;
+                                        }
+                                    }
+                                    return false;
+                                }
+                            })
+                            .setResteasyReactiveRecorder(recorder)
+                            .setApplicationClassPredicate(applicationClassPredicate);
 
             if (!serverDefaultProducesHandlers.isEmpty()) {
                 List<DefaultProducesHandler> handlers = new ArrayList<>(serverDefaultProducesHandlers.size());

--- a/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
+++ b/independent-projects/resteasy-reactive/common/processor/src/main/java/org/jboss/resteasy/reactive/common/processor/EndpointIndexer.java
@@ -715,6 +715,7 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
             ParameterType type, String elementType, boolean single, String signature);
 
     private String[] applyDefaultProduces(String[] produces, Type nonAsyncReturnType) {
+        setupApplyDefaults(nonAsyncReturnType);
         if (produces != null && produces.length != 0)
             return produces;
         return applyAdditionalDefaults(nonAsyncReturnType);
@@ -734,6 +735,10 @@ public abstract class EndpointIndexer<T extends EndpointIndexer<T, PARAM, METHOD
             }
         }
         return result.toArray(EMPTY_STRING_ARRAY);
+    }
+
+    protected void setupApplyDefaults(Type nonAsyncReturnType) {
+
     }
 
     protected String[] applyAdditionalDefaults(Type nonAsyncReturnType) {

